### PR TITLE
cam6_3_050: New CAM CARMA tests

### PR DESCRIPTION
**Pull request by manderberg**
Adding new CARMA tests to CAM regression testing.
Adding a new CARMA testing category.
Issue not closing; more tests coming in another PR.

Addresses #420

### CODE REVIEW

**Review by fvitt — APPROVED** [review]
Looks OK to me.  Thanks for the work on this.

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands [hunk 1]
```diff
@@ -0,0 +1,4 @@
+./xmlchange CAM_CONFIG_OPTS=" -phys cam4 -chem waccm_ma_sulfur -carma mixed_sulfate"
+./xmlchange ROF_NCPL=\$ATM_NCPL
+./xmlchange GLC_NCPL=\$ATM_NCPL
+./xmlchange RUN_STARTDATE="0001-01-01"
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands:4]: I do not see a test using this file, please remove.

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cam [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+inithist='ENDOFRUN'
+pbuf_global_allocate=.false.
+history_carma=.true.
+carma_do_fixedinit=.false.
+carma_reftfile="camrun.cam.r.carma_reft.nc"
+flbc_list              = 'CCL4', 'CF2CLBR', 'CF3BR', 'CFC11', 'CFC113', 'CFC12', 'CH3BR', 'CH3CCL3', 'CH3CL', 'CH4', 'CO2', 'H2',
+         'HCFC22', 'N2O', 'OCS' 
+solar_data_type='FIXED'
+solar_data_ymd=20000101
+
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cam:13]: I do not see a test using this file, please remove.

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_clm [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+!                        (includes $inst_string for multi-ensemble cases)
+! Set glc_grid           with CISM_GRID                          option
+! Set glc_smb            with GLC_SMB                            option
+! Set maxpatch_glcmec    with GLC_NEC                            option
+! Set glc_do_dynglacier  with GLC_TWO_WAY_COUPLING               env variable
+!----------------------------------------------------------------------------------
+hist_nhtfrq = 3
+hist_mfilt  = 1
+hist_ndens  = 1
+
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_clm:27]: I do not see a test using this file, please remove.

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cpl [hunk 4]
```diff
@@ -0,0 +1,2 @@
+reprosum_diffmax=1.0e-14
+reprosum_recompute=.true.
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cpl:2]: I do not see a test using this file, please remove.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands [hunk 5]
```diff
@@ -0,0 +1,2 @@
+./xmlchange CAM_CONFIG_OPTS="-phys cam5 -aquaplanet -carma test_detrain"
+./xmlchange RUN_STARTDATE="0001-01-01"
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands:2]: I do not see a test using this file, please remove.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/user_nl_cam [hunk 6]
```diff
@@ -0,0 +1,9 @@
+mfilt=1,1,1,1,1,1
+ndens=1,1,1,1,1,1
+nhtfrq=3,3,3,3,3,3
+inithist='ENDOFRUN'
+pbuf_global_allocate=.false.
+history_carma=.true.
+carma_do_fixedinit=.false.
+carma_do_cldice=.true.
+carma_reftfile="camrun.cam.r.carma_reft.nc"
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/user_nl_cam:9]: I do not see a test using this file, please remove.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/user_nl_cpl [hunk 7]
```diff
@@ -0,0 +1,2 @@
+reprosum_diffmax=1.0e-14
+reprosum_recompute=.true.
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/user_nl_cpl:2]: I do not see a test using this file, please remove.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_radiative/user_nl_clm [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+!                        (includes $inst_string for multi-ensemble cases)
+! Set glc_grid           with CISM_GRID                          option
+! Set glc_smb            with GLC_SMB                            option
+! Set maxpatch_glcmec    with GLC_NEC                            option
+! Set glc_do_dynglacier  with GLC_TWO_WAY_COUPLING               env variable
+!----------------------------------------------------------------------------------
+hist_nhtfrq = 3
+hist_mfilt  = 1
+hist_ndens  = 1
+
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_test_radiative/user_nl_clm:27]: This test is for an aquaplanet compset so this file is not used. Please remove.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_tracers2/user_nl_clm [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+!                        (includes $inst_string for multi-ensemble cases)
+! Set glc_grid           with CISM_GRID                          option
+! Set glc_smb            with GLC_SMB                            option
+! Set maxpatch_glcmec    with GLC_NEC                            option
+! Set glc_do_dynglacier  with GLC_TWO_WAY_COUPLING               env variable
+!----------------------------------------------------------------------------------
+hist_nhtfrq = 3
+hist_mfilt  = 1
+hist_ndens  = 1
+
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/carma_test_tracers2/user_nl_clm:27]: This test is for an aquaplanet compset so this file is not used. Please remove.

File: cime_config/testdefs/testmods_dirs/cam/outfrq3s_carma/shell_commands [hunk 10]
```diff
@@ -1,5 +1,4 @@
 ./xmlchange --append CAM_CONFIG_OPTS="-carma test_tracers"
 ./xmlchange ROF_NCPL=\$ATM_NCPL
 ./xmlchange GLC_NCPL=\$ATM_NCPL
-./xmlchange CAM_NML_USE_CASE=UNSET
-./xmlchange RUN_STARTDATE="00000101"
+./xmlchange RUN_STARTDATE="0001-01-01"
```
Comment (gold2718) [inline cime_config/testdefs/testmods_dirs/cam/outfrq3s_carma/shell_commands:4]: It looks like this file is no longer used. Please remove.

**Review by gold2718 — CHANGES_REQUESTED** [review]
I think there are a few files that are not being used and should be removed.
Otherwise, it looks okay.

**Comment by fvitt** [comment]
Should we remove the carma models that do not have regression tests?  
These models do not have tests:
```
 bc_strat
 cirrus
 cirrus_dust
 pmc_sulfate
 tholin
```
@cbardeen , @tilmes thoughts?

**Comment by cbardeen** [comment]
I don’t particularly care about bcstrat or tholin, but do want the other ones retained. Brian Toon’s group is working on enhancing the cirrus model. Mike did publish some papers with bcstrat, but I think we have better soot models now.

**Comment by tilmes** [comment]
I think, we agreed to remove tholin, pmc_sulfate,  bc_strat, and cirrus_dust (if I recall). I think, cirrus  is a model we did not get to run, but want to leave in, is that correct?

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands (hunk 1 above)
Comment (fvitt) [reply cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands:4]: This is a test we would like to keep.  Currently it fails.  We will try to fix this test.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands (hunk 5 above)
Comment (fvitt) [reply cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands:2]: We would like to keep this test.  Currently fails with DEBUG=TRUE.

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands [hunk 11]
```diff
@@ -0,0 +1,4 @@
+./xmlchange CAM_CONFIG_OPTS=" -phys cam4 -chem waccm_ma_sulfur -carma mixed_sulfate"
```
Comment (fvitt) [inline cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands:1]:
```suggestion
./xmlchange CAM_CONFIG_OPTS=" -phys cam4 -rad rrtmg -chem waccm_ma_sulfur -carma mixed_sulfate"
```
@manderberg : Try this change for mixed_sulfate.  Add this back to the the test list if this works.

**Review by fvitt — CHANGES_REQUESTED** [review]

File: cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands (hunk 5 above)
Comment (gold2718) [reply cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands:2]: What is the test that is failing? It might be related to a known bug.

Comment (fvitt) [reply cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands:2]: @gold2718 This is the failing test:
`ERC_D_Ln9_Vmct.f10_f10_mg37.QPC5.cheyenne_intel.cam-carma_test_detrain`
I think it is an issue with test_detrain CARMA model.

File: cime_config/testdefs/testmods_dirs/cam/carma_meteor_impact/user_nl_clm [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+!                        (includes $inst_string for multi-ensemble cases)
+! Set glc_grid           with CISM_GRID                          option
+! Set glc_smb            with GLC_SMB                            option
+! Set maxpatch_glcmec    with GLC_NEC                            option
+! Set glc_do_dynglacier  with GLC_TWO_WAY_COUPLING               env variable
+!----------------------------------------------------------------------------------
+hist_nhtfrq = 3
+hist_mfilt  = 1
+hist_ndens  = 1
+
```
[the remark this answers is not in the document]
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_meteor_impact/user_nl_clm:27]: Done.

File: cime_config/testdefs/testmods_dirs/cam/carma_meteor_smoke/user_nl_clm [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+!                        (includes $inst_string for multi-ensemble cases)
+! Set glc_grid           with CISM_GRID                          option
+! Set glc_smb            with GLC_SMB                            option
+! Set maxpatch_glcmec    with GLC_NEC                            option
+! Set glc_do_dynglacier  with GLC_TWO_WAY_COUPLING               env variable
+!----------------------------------------------------------------------------------
+hist_nhtfrq = 3
+hist_mfilt  = 1
+hist_ndens  = 1
+
```
[the remark this answers is not in the document]
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_meteor_smoke/user_nl_clm:27]: Done.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_radiative/user_nl_clm (hunk 8 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_test_radiative/user_nl_clm:27]: Done.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_tracers2/user_nl_clm (hunk 9 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_test_tracers2/user_nl_clm:27]: Done.

File: cime_config/testdefs/testmods_dirs/cam/outfrq3s_carma/shell_commands (hunk 10 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/outfrq3s_carma/shell_commands:4]: Directory cime_config/testdefs/testmods_dirs/cam/outfrq3s_carma is no longer needed so it has been removed.

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands (hunk 11 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands:1]: tholin, pmc_sulfate and bc_strat were not included in any of my testing , and aren't in testlist_cam.xml. 
I was never able to get cirrus or cirrus_dust to run; neither it nor cirrus are in testlist_cam.xml.
With the addition of "-rad rrtmg" to CAM_CONFIG_OPTS mixed_sulfate passes the SMS test on cheyenne and on izumi. Adding mixed_sulfate back in to testing.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands (hunk 5 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands:2]: From cesm.log.2015979.chadmin1.ib0.cheyenne.ucar.edu.211215-144033 (case ERC_D_Ln9_Vmct.f10_f10_mg37.QPC5.cheyenne_intel.cam-carma_test_detrain.20121215-03):

"ERROR: Attempt to find undefined name in pbuf TND_QSNOW
16:Image              PC                Routine            Line        Source             
16:cesm.exe           000000000558A646  Unknown               Unknown  Unknown
16:cesm.exe           0000000004987062  shr_abort_mod_mp_         114  shr_abort_mod.F90
16:cesm.exe           0000000004986EC8  shr_abort_mod_mp_          61  shr_abort_mod.F90
16:cesm.exe           00000000010377B8  physics_buffer_mp         807  physics_buffer.F90.in
16:cesm.exe           000000000288BFB1  carma_intr_mp_car        1083  carma_intr.F90
16:cesm.exe           00000000011F3419  physpkg_mp_tphysb        2382  physpkg.F90
16:cesm.exe           00000000011CE1D3  physpkg_mp_phys_r        1155  physpkg.F90
16:cesm.exe           0000000000888688  cam_comp_mp_cam_r         258  cam_comp.F90
16:cesm.exe           0000000000846150  atm_comp_mct_mp_a         355  atm_comp_mct.F90
16:cesm.exe           00000000004761F3  component_mod_mp_         257  component_mod.F90
16:cesm.exe           000000000042E947  cime_comp_mod_mp_        2281  cime_comp_mod.F90
16:cesm.exe           000000000046CD0D  MAIN__                    122  cime_driver.F90"

Doing "grep -Hnr TND_QSNOW" from <CAM root>/src produces:
"physics/carma/cam/carma_intr.F90:1083:      call pbuf_get_field(pbuf, pbuf_get_index("TND_QSNOW"), tnd_qsnow)
physics/cam/micro_mg_cam.F90:796:      call pbuf_add_field('TND_QSNOW',  'physpkg',dtype_r8,(/pcols,pver/), tnd_qsnow_idx)
control/cam_snapshot.F90:1716:          'TND_QSNOW              ','unset                  ',&"

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cam (hunk 2 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cam:13]: Problems with mixed_sulfate have been resolved; it now passed the ERS test on both cheyenne and izumi. The file testlist_cam.xml has been updated to reflect this.

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cpl (hunk 4 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/user_nl_cpl:2]: Problems resolved, mixed_sulfate passes ERS test on cheyenne and izumi.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands (hunk 5 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/shell_commands]: I have removed the test_detrain sub-directory from <root>/cime_config/testdefs/testmods_dirs/cam.

File: cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/user_nl_cam (hunk 6 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_test_detrain/user_nl_cam]: Done.

File: cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands (hunk 1 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_mixed_sulfate/shell_commands]: Problem is resolved, mixed_sulfate passed the ERS test on cheyenne and izumi.

**Review by fvitt — APPROVED** [review]

**Review by gold2718 — APPROVED** [review]
Okay, looks good now, thanks!
I will get this in the PR pipeline.

File: cime_config/config_compsets.xml [hunk 14]
```diff
@@ -446,6 +446,10 @@
     <lname>HIST_CAM60%WCTS%SDYN_CLM50%SP_CICE%PRES_DOCN%DOM_MOSART_SGLC_SWAV</lname>
   </compset>
 
+  <compset>
+    <alias>FW4ma2000</alias>
+    <lname>2000_CAM40%WCCM_CLM50%SP_CICE%PRES_DOCN%DOM_RTM_SGLC_SWAV</lname>
```
Comment (cacraigucar) [inline cime_config/config_compsets.xml:451]: Just an observation - I find it interesting that we are introducing a CAM4 compset as there have been rumblings about trying to phase out older versions of CAM

File: cime_config/testdefs/testlist_cam.xml [hunk 15]
```diff
@@ -610,13 +610,147 @@
...
+      <option name="comment" >CARMA dust test</option>
+      <option name="wallclock">00:30:00</option>
+    </options>
+  </test>
+<!--              carma sea_salt          -->
+  <test compset="QPC5" grid="f10_f10_mg37" name="ERC_D_Ln9_Vmct" testmods="cam/carma_sea_salt">
+    <machines>
+      <machine name="izumi" compiler="nag" category="carma"/>
       <machine name="izumi" compiler="nag" category="aux_cam"/>
+      <machine name="cheyenne" compiler="intel" category="carma"/>
```
Comment (cacraigucar) [inline cime_config/testdefs/testlist_cam.xml:629]: As CARMA becomes more mainstream, you might want to consider adding a prealpha or prebeta CARMA test.  This doesn't need to happen now, but is something to keep in mind.  Checking in prealpha or prebeta has the advantage of testing in the CESM checkout while aux_cam tests a CAM checkout.  A single test (probably at prebeta) should be sufficient to make sure that CARMA is functioning in a CESM checkout.

**Review by cacraigucar — APPROVED** [review]

**Review by peverwhee — APPROVED** [review]

File: cime_config/testdefs/testmods_dirs/cam/carma_meteor_impact/shell_commands [hunk 16]
```diff
@@ -0,0 +1,5 @@
+./xmlchange --append CAM_CONFIG_OPTS="-carma meteor_impact"
+./xmlchange --append CAM_CONFIG_OPTS="-max_n_rad_cnst=100"
+./xmlchange ROF_NCPL=\$ATM_NCPL
+./xmlchange GLC_NCPL=\$ATM_NCPL
```
Comment (nusbaume) [inline cime_config/testdefs/testmods_dirs/cam/carma_meteor_impact/shell_commands:4]: Do these `xmlchange` commands need to be here?  It looks like this is an aquaplanet simulation (i.e. ROF and GLC are just stubs).

I have the same question for the "carma_meteor_smoke" and "test_tracers2" `shell_commands` files as well.

**Review by nusbaume — APPROVED** [review]
Looks good to me!  Just one possible clean-up related question.

File: cime_config/testdefs/testmods_dirs/cam/carma_meteor_impact/shell_commands (hunk 16 above)
Comment (manderberg) [reply cime_config/testdefs/testmods_dirs/cam/carma_meteor_impact/shell_commands]: I have removed these two lines from the shell_commands file for the carma_meteor_impact, carma_meteor_smoke and carma_test_tracers2 models. The changes have been committed and pushed up to my fork.

File: cime_config/config_compsets.xml (hunk 14 above)
Comment (gold2718) [reply cime_config/config_compsets.xml:451]: Maybe so but both the simpler modeling community and the paleo community view CAM4 as the only stable physics system that is affordable for longer (e.g., millennium) simulations. WACCM can also be expensive so I feel we will be supporting CAM4 longer than CAM3 or CAM5.

Comment (fvitt) [reply cime_config/config_compsets.xml:451]: At least one of the CARMA models needs CAM4 physics which is why we added it.

**Pull request merged**